[pytorch] Fix mkldnn heuristic for multithreaded convolution

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52909 [pytorch] Fix mkldnn heuristic for multithreaded convolution**

PR #46675 introduced heuristics to use thnn_conv2d for 1x1
convolutions, since mkldnn had a bug that was slowing those cases
down. Unfortunately, the test plan for that PR only tested single-threaded
convolutions; mkldnn is considerably faster on multithreaded convolutions.

An example from yolov3, on 24 cores of a Xeon Platinum 8175M CPU @ 2.50GHz
```
input:{1, 64, 192, 256}, weight:{32, 64, 1, 1}
thnn_conv2d: GFLOPS/s=104.574G/s
mkldnn_convolution: GFLOPS/s=467.357G/s
```

Differential Revision: [D26685272](https://our.internmc.facebook.com/intern/diff/D26685272/)

bertmaher committed:
[pytorch] Fix mkldnn heuristic for multithreaded convolution

PR #46675 introduced heuristics to use thnn_conv2d for 1x1
convolutions, since mkldnn had a bug that was slowing those cases
down. Unfortunately, the test plan for that PR only tested single-threaded
convolutions; mkldnn is considerably faster on multithreaded convolutions.

An example from yolov3, on 24 cores of a Xeon Platinum 8175M CPU @ 2.50GHz
```
input:{1, 64, 192, 256}, weight:{32, 64, 1, 1}
thnn_conv2d: GFLOPS/s=104.574G/s
mkldnn_convolution: GFLOPS/s=467.357G/s
```

Differential Revision: [D26685272](https://our.internmc.facebook.com/intern/diff/D26685272/)

[ghstack-poisoned]

xuzhao9 commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -236,15 +236,15 @@ auto ConvParams::use_mkldnn(const at::Tensor& input, const at::Tensor& weight) c
      !transposed && // or transposed tensors
      (is_strided() || is_dilated() || input.size(0) >= 16 ||
-      weight.size(-1) != 1 || weight.size(-2) != 1) &&
+      weight.size(-1) != 1 || weight.size(-2) != 1 || at::get_num_threads() > 1) &&
```
Would it be helpful to add a comment to explain why adding this heuristic?

bertmaher committed:
Add the missing Parallel.h header on "[pytorch] Fix mkldnn heuristic for multithreaded convolution"


PR #46675 introduced heuristics to use thnn_conv2d for 1x1
convolutions, since mkldnn had a bug that was slowing those cases
down. Unfortunately, the test plan for that PR only tested single-threaded
convolutions; mkldnn is considerably faster on multithreaded convolutions.

An example from yolov3, on 24 cores of a Xeon Platinum 8175M CPU @ 2.50GHz
```
input:{1, 64, 192, 256}, weight:{32, 64, 1, 1}
thnn_conv2d: GFLOPS/s=104.574G/s
mkldnn_convolution: GFLOPS/s=467.357G/s
```

Differential Revision: [D26685272](https://our.internmc.facebook.com/intern/diff/D26685272/)

[ghstack-poisoned]

wconstab reviewed (approved):
lgtm

did you have a way to verify this fixes the regression, or do you need help measuring? Wish we had the A/B PR thing working by now, but one step at a time..

bertmaher commented:
> lgtm
> 
> did you have a way to verify this fixes the regression, or do you need help measuring? Wish we had the A/B PR thing working by now, but one step at a time..

I spot checked yolov3 on of our AWS benchmarking machines, and it reclaims the 10% loss in the linked task.  I also ran these through a convolution benchmark harness that I've been developing on the side and verified that this gives significant wins for multithreaded 1x1 convs.
